sydent/hs_federation/verifier.py

```diff
@@ -0,0 +1,206 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import time
+
+from twisted.internet import defer
+from twisted.names.error import DNSNameError
+import twisted.names.client
+import twisted.names.dns
+from unpaddedbase64 import decode_base64
+import signedjson.sign
+import signedjson.key
+from signedjson.sign import SignatureVerifyException
+
+from sydent.http.httpclient import FederationHttpClient
+
+
+logger = logging.getLogger(__name__)
+
+
+class NoAuthenticationError(Exception):
+    pass
+
+
+"""
+Verifies signed json blobs from Matrix Homeservers by finding the
+Homeserver's address, contacting it, requesting its keys and
+verifying that the signature on the json blob matches.
+"""
+class Verifier(object):
+    def __init__(self, sydent):
+        self.sydent = sydent
+        # Cache of server keys. These are cached until the 'valid_until_ts' time
+        # in the result.
+        self.cache = {
+            # server_name: <result from keys query>,
+        }
+
+    @defer.inlineCallbacks
+    def _getEndpointForServer(self, server_name):
+        if ':' in server_name:
+            defer.returnValue(tuple(server_name.rsplit(':', 1)))
+
+        service_name = "%s.%s.%s" % ('_matrix', '_tcp', server_name)
+
+        default = server_name, 8448
+
+        try:
+            answers, _, _ = yield twisted.names.client.lookupService(service_name)
+        except DNSNameError:
+            logger.info("DNSNameError doing SRV lookup for %s - using default", service_name)
+            defer.returnValue(default)
+
+        for answer in answers:
+            if answer.type != twisted.names.dns.SRV or not answer.payload:
+                continue
+
+            # XXX we just use the first
+            logger.info("Got SRV answer: %r / %d for %s", str(answer.payload.target), answer.payload.port, service_name)
+            defer.returnValue((str(answer.payload.target), answer.payload.port))
+
+        logger.info("No valid answers found in response from %s (%r)", server_name, answers)
+        defer.returnValue(default)
+
+    @defer.inlineCallbacks
+    def _getKeysForServer(self, server_name):
+        """Get the signing key data from a home server.
+        """
+
+        if server_name in self.cache:
+            cached = self.cache[server_name]
+            now = int(time.time() * 1000)
+            if cached['valid_until_ts'] > now:
+                defer.returnValue(self.cache[server_name]['verify_keys'])
+
+        host_port = yield self._getEndpointForServer(server_name)
+        logger.info("Got host/port %s/%s for %s", host_port[0], host_port[1], server_name)
+        client = FederationHttpClient(self.sydent)
+        result = yield client.get_json("https://%s:%s/_matrix/key/v2/server/" % host_port)
+        if 'verify_keys' not in result:
+            raise Exception("No key found in response")
```

Comment:
shouldn't this be a `SignatureVerifyException` ?

Reply:
yes, it probably ought to